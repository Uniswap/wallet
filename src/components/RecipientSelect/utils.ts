import { SectionListData } from 'react-native'
import { SearchableRecipient } from 'src/components/RecipientSelect/types'
import { unique } from 'src/utils/array'

interface HasAddress {
  address: string
}

export function uniqueAddressesOnly<T extends HasAddress>(objectsWithAddress: T[]): T[] {
  // the input array must be objects that have an obj.address field
  // had to cast to any because ts doesn't recognize it as HasAddress... maybe issue with unique
  return unique(
    objectsWithAddress,
    (v, i, a) => a.findIndex((v2) => v2.address === v.address) === i
  )
}

export function filterSections(
  sections: SectionListData<SearchableRecipient>[],
  filteredAddresses: string[]
): { title: string; data: SearchableRecipient[] }[] {
  return sections
    .map((section) => {
      const { title, data } = section
      return { title, data: data.filter((item) => filteredAddresses.includes(item.address)) }
    })
    .filter((section) => section.data.length > 0)
}
