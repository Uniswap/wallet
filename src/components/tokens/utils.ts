import { TFunction } from 'i18next'
import { SafetyLevel } from 'src/data/__generated__/types-and-hooks'

export function getTokenSafetyHeaderText(
  safetyLevel: NullUndefined<SafetyLevel>,
  t: TFunction
): string | undefined {
  switch (safetyLevel) {
    case SafetyLevel.MediumWarning:
      return t('Caution')
    case SafetyLevel.StrongWarning:
      return t('Warning')
    case SafetyLevel.Blocked:
      return t('Not available')
  }
}
