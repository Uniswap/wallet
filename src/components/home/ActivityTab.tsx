import { NetworkStatus } from '@apollo/client'
import { FlashList } from '@shopify/flash-list'
import { TFunction } from 'i18next'
import React, { forwardRef, useCallback, useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import { NativeScrollEvent, NativeSyntheticEvent } from 'react-native'
import { useAppSelector } from 'src/app/hooks'
import { NoTransactions } from 'src/components/icons/NoTransactions'
import { Box, Flex } from 'src/components/layout'
import { AnimatedFlashList } from 'src/components/layout/AnimatedFlashList'
import { BaseCard } from 'src/components/layout/BaseCard'
import { TabContentProps, TAB_STYLES } from 'src/components/layout/TabHelpers'
import { Loader } from 'src/components/loading'
import { Text } from 'src/components/Text'
import { EMPTY_ARRAY, PollingInterval } from 'src/constants/misc'
import { isNonPollingRequestInFlight } from 'src/data/utils'
import { useTransactionListQuery } from 'src/data/__generated__/types-and-hooks'
import { usePersistedError } from 'src/features/dataApi/utils'
import {
  formatTransactionsByDate,
  parseDataResponseToTransactionDetails,
} from 'src/features/transactions/history/utils'
import { useMergeLocalAndRemoteTransactions } from 'src/features/transactions/hooks'
import TransactionSummaryRouter from 'src/features/transactions/SummaryCards/TransactionSummaryRouter'
import { TransactionDetails } from 'src/features/transactions/types'
import { AccountType } from 'src/features/wallet/accounts/types'
import { useActiveAccountWithThrow } from 'src/features/wallet/hooks'
import { makeSelectAccountHideSpamTokens } from 'src/features/wallet/selectors'
import { usePollOnFocusOnly } from 'src/utils/hooks'

type ActivityTabProps = {
  owner: string
  containerProps?: TabContentProps
  scrollHandler?: (event: NativeSyntheticEvent<NativeScrollEvent>) => void
}

const PENDING_TITLE = (t: TFunction): string => t('Pending')
const TODAY_TITLE = (t: TFunction): string => t('Today')
const MONTH_TITLE = (t: TFunction): string => t('This Month')

type LoadingItem = {
  itemType: 'LOADING'
  id: number
}
const isLoadingItem = (x: TransactionDetails | SectionHeader | LoadingItem): x is LoadingItem =>
  'itemType' in x && x.itemType === 'LOADING'

type SectionHeader = {
  itemType: 'HEADER'
  title: string
}
const isSectionHeader = (x: TransactionDetails | SectionHeader | LoadingItem): x is SectionHeader =>
  'itemType' in x && x.itemType === 'HEADER'

const LOADING_ITEM = (index: number): LoadingItem => ({ itemType: 'LOADING', id: index })
const LOADING_DATA = [LOADING_ITEM(1), LOADING_ITEM(2), LOADING_ITEM(3), LOADING_ITEM(4)]

const FOOTER_HEIGHT = 20
const ESTIMATED_ITEM_SIZE = 92

const SectionTitle = ({ title }: { title: string }): JSX.Element => (
  <Box pb="spacing2" pt="spacing12">
    <Text color="textSecondary" variant="subheadSmall">
      {title}
    </Text>
  </Box>
)

const renderActivityItem = (
  item: TransactionDetails | SectionHeader | LoadingItem,
  readonly: boolean
): JSX.Element => {
  // if it's a loading item, render the loading placeholder
  if (isLoadingItem(item)) {
    return <Loader.Transaction />
  }
  // if it's a section header, render it differently
  if (isSectionHeader(item)) {
    return <SectionTitle title={item.title} />
  }
  return (
    <TransactionSummaryRouter
      readonly={readonly}
      // TODO: [MOB-3879] @ianlapham Implement finalized inline/card designs for special case txns (failure, pending failure, etc)
      // When finalized, delete this prop if no longer using inline copy.
      showInlineWarning={false}
      transaction={item}
    />
  )
}

export const ActivityTab = forwardRef<FlashList<unknown>, ActivityTabProps>(
  ({ owner, containerProps, scrollHandler }, ref) => {
    const { t } = useTranslation()
    const { type } = useActiveAccountWithThrow()
    const readonly = type === AccountType.Readonly

    const keyExtractor = useCallback(
      (info: TransactionDetails | SectionHeader | LoadingItem) => {
        // for loading items, use the index as the key
        if (isLoadingItem(info)) {
          return `${owner}-${info.id}`
        }
        // for section headers, use the title as the key
        if (isSectionHeader(info)) {
          return `${owner}-${info.title}`
        }
        // for transactions, use the transaction hash as the key
        return info.id
      },
      [owner]
    )

    const {
      refetch,
      networkStatus,
      loading: requestLoading,
      data,
      error: requestError,
      startPolling,
      stopPolling,
    } = useTransactionListQuery({
      variables: { address: owner },
      notifyOnNetworkStatusChange: true,
    })

    usePollOnFocusOnly(startPolling, stopPolling, PollingInterval.Fast)

    // Hide all spam transactions if active wallet has enabled setting.
    const activeAccount = useActiveAccountWithThrow()
    const hideSpamTokens = useAppSelector<boolean>(
      makeSelectAccountHideSpamTokens(activeAccount.address)
    )

    const formattedTransactions = useMemo(() => {
      if (!data) return EMPTY_ARRAY

      const parsedTxHistory = parseDataResponseToTransactionDetails(data, hideSpamTokens)

      return parsedTxHistory
    }, [data, hideSpamTokens])

    const transactions = useMergeLocalAndRemoteTransactions(owner, formattedTransactions)

    // Format transactions for section list
    const {
      pending,
      todayTransactionList,
      monthTransactionList,
      priorByMonthTransactionList,
      combinedTransactionList,
    } = useMemo(() => formatTransactionsByDate(transactions), [transactions])

    const hasTransactions = combinedTransactionList?.length > 0

    const hasData = !!data?.portfolios?.[0]?.assetActivities
    const isLoading = isNonPollingRequestInFlight(networkStatus)
    const isError = usePersistedError(requestLoading, requestError)

    // show loading if no data and fetching, or refetching when there is error (for UX when "retry" is clicked).
    const showLoading =
      (!hasData && isLoading) || (Boolean(isError) && networkStatus === NetworkStatus.refetch)

    const sectionData = useMemo(() => {
      if (showLoading) {
        return LOADING_DATA
      }

      if (!hasTransactions) {
        return EMPTY_ARRAY
      }
      return [
        ...(pending.length > 0 ? [PENDING_TITLE(t), ...pending] : EMPTY_ARRAY),
        ...(todayTransactionList.length > 0
          ? [{ itemType: 'HEADER', title: TODAY_TITLE(t) }, ...todayTransactionList]
          : EMPTY_ARRAY),
        ...(monthTransactionList.length > 0
          ? [{ itemType: 'HEADER', title: MONTH_TITLE(t) }, ...monthTransactionList]
          : EMPTY_ARRAY),
        // for each month prior, detect length and render if includes transactions
        ...Object.keys(priorByMonthTransactionList).reduce(
          (accum: (TransactionDetails | SectionHeader | LoadingItem)[], month) => {
            const transactionList = priorByMonthTransactionList[month]
            if (transactionList && transactionList.length > 0) {
              accum.push({ itemType: 'HEADER', title: month }, ...transactionList)
            }
            return accum
          },
          []
        ),
      ]
    }, [
      hasTransactions,
      monthTransactionList,
      pending,
      priorByMonthTransactionList,
      t,
      todayTransactionList,
      showLoading,
    ])

    /**
     * If tab container is smaller than the approximate screen height, we need to manually add
     * padding so scroll works as intended since minHeight is not supported by FlashList in
     * `contentContainerStyle`. Padding is proportional to the number of rows the data items take up.
     */
    const footerPadding =
      transactions.length < 6
        ? (ESTIMATED_ITEM_SIZE * (6 - transactions.length)) / 2
        : FOOTER_HEIGHT

    const onRetry = useCallback(() => {
      refetch({
        address: owner,
      })
    }, [owner, refetch])

    if (!hasData && isError) {
      return (
        <Box height="100%" pb="spacing60">
          <BaseCard.ErrorState
            retryButtonLabel={t('Retry')}
            title={t('Couldn’t load activity')}
            onRetry={onRetry}
          />
        </Box>
      )
    }

    return transactions.length === 0 ? (
      <Flex centered grow height="100%">
        <BaseCard.EmptyState
          description={t('When this wallet makes transactions, they’ll appear here.')}
          icon={<NoTransactions />}
          title={t('No activity yet')}
        />
      </Flex>
    ) : (
      <Flex grow style={TAB_STYLES.tabListContainer}>
        <AnimatedFlashList
          ref={ref}
          ListFooterComponent={
            // If not loading, we add a footer  to cover any possible space that is covered up by bottom tab bar
            networkStatus === NetworkStatus.fetchMore ? (
              <Box p="spacing12">
                <Loader.Transaction repeat={4} />
              </Box>
            ) : (
              <Box height={footerPadding} />
            )
          }
          data={sectionData}
          estimatedItemSize={ESTIMATED_ITEM_SIZE}
          keyExtractor={keyExtractor}
          numColumns={1}
          renderItem={({ item }): JSX.Element => renderActivityItem(item, readonly)}
          showsVerticalScrollIndicator={false}
          onScroll={scrollHandler}
          {...containerProps}
        />
      </Flex>
    )
  }
)
