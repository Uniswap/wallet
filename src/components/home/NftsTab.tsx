import { NetworkStatus } from '@apollo/client'
import { FlashList } from '@shopify/flash-list'
import { ImpactFeedbackStyle } from 'expo-haptics'
import React, { forwardRef, useCallback } from 'react'
import { useTranslation } from 'react-i18next'
import { ListRenderItemInfo, NativeScrollEvent, NativeSyntheticEvent, View } from 'react-native'
import { useAppDispatch, useAppTheme } from 'src/app/hooks'
import { useAppStackNavigation } from 'src/app/navigation/types'
import NoNFTsIcon from 'src/assets/icons/empty-state-picture.svg'
import { TouchableArea } from 'src/components/buttons/TouchableArea'
import { NFTViewer } from 'src/components/images/NFTViewer'
import { AnimatedFlashList } from 'src/components/layout/AnimatedFlashList'
import { BaseCard } from 'src/components/layout/BaseCard'
import { Box } from 'src/components/layout/Box'
import { Flex } from 'src/components/layout/Flex'
import { TabContentProps } from 'src/components/layout/TabHelpers'
import { Loader } from 'src/components/loading'
import { ScannerModalState } from 'src/components/QRCodeScanner/constants'
import { EMPTY_ARRAY } from 'src/constants/misc'
import { isError, isNonPollingRequestInFlight } from 'src/data/utils'
import { NftsTabQuery, useNftsTabQuery } from 'src/data/__generated__/types-and-hooks'
import { openModal } from 'src/features/modals/modalSlice'
import { NFTItem } from 'src/features/nfts/types'
import { getNFTAssetKey } from 'src/features/nfts/utils'
import { ModalName } from 'src/features/telemetry/constants'
import { removePendingSession } from 'src/features/walletConnect/walletConnectSlice'
import { Screens } from 'src/screens/Screens'

const MAX_NFT_IMAGE_SIZE = 375
const ESTIMATED_ITEM_SIZE = 251 // heuristic provided by FlashList
const PREFETCH_ITEMS_THRESHOLD = 0.5
const LOADING_ITEM = 'loading'
const FOOTER_HEIGHT = 20

type NftsTabProps = {
  owner: string
  containerProps?: TabContentProps
  scrollHandler?: (event: NativeSyntheticEvent<NativeScrollEvent>) => void
}

function formatNftItems(data: NftsTabQuery | undefined): NFTItem[] {
  const items = data?.nftBalances?.edges?.flatMap((item) => item.node)
  if (!items) return EMPTY_ARRAY
  const nfts = items
    .filter((item) => item?.ownedAsset?.nftContract?.address && item?.ownedAsset?.tokenId)
    .map((item): NFTItem => {
      return {
        name: item?.ownedAsset?.name ?? undefined,
        description: item?.ownedAsset?.description ?? undefined,
        contractAddress: item?.ownedAsset?.nftContract?.address ?? undefined,
        tokenId: item?.ownedAsset?.tokenId ?? undefined,
        imageUrl: item?.ownedAsset?.image?.url ?? undefined,
        collectionName: item?.ownedAsset?.collection?.name ?? undefined,
        isVerifiedCollection: item?.ownedAsset?.collection?.isVerified ?? undefined,
        floorPrice: item?.ownedAsset?.collection?.markets?.[0]?.floorPrice?.value ?? undefined,
        imageDimensions:
          item?.ownedAsset?.image?.dimensions?.height && item?.ownedAsset?.image?.dimensions?.width
            ? {
                width: item?.ownedAsset?.image.dimensions.width,
                height: item?.ownedAsset?.image.dimensions.height,
              }
            : undefined,
      }
    })
  return nfts
}

const keyExtractor = (item: NFTItem | string): string =>
  typeof item === 'string'
    ? LOADING_ITEM
    : getNFTAssetKey(item.contractAddress ?? '', item.tokenId ?? '')

export const NftsTab = forwardRef<FlashList<unknown>, NftsTabProps>(
  ({ owner, containerProps, scrollHandler }, ref) => {
    const navigation = useAppStackNavigation()
    const { t } = useTranslation()
    const theme = useAppTheme()
    const dispatch = useAppDispatch()

    const { data, fetchMore, refetch, networkStatus } = useNftsTabQuery({
      variables: { ownerAddress: owner, first: 30 },
      notifyOnNetworkStatusChange: true, // Used to trigger network state / loading on refetch or fetchMore
      errorPolicy: 'all', // Suppress non-null image.url fields from backend
    })

    const nftDataItems = formatNftItems(data)
    const shouldAddInLoadingItem =
      networkStatus === NetworkStatus.fetchMore && nftDataItems.length % 2 === 1

    const onListEndReached = useCallback(() => {
      if (!data?.nftBalances?.pageInfo?.hasNextPage) return

      fetchMore({
        variables: {
          first: 30,
          after: data?.nftBalances?.pageInfo?.endCursor,
        },
      })
    }, [
      data?.nftBalances?.pageInfo?.endCursor,
      data?.nftBalances?.pageInfo?.hasNextPage,
      fetchMore,
    ])

    const onPressItem = useCallback(
      (asset: NFTItem) => {
        // Always ensure push to enforce right push
        navigation.push(Screens.NFTItem, {
          owner,
          address: asset.contractAddress ?? '',
          tokenId: asset.tokenId ?? '',
          collectionName: asset.collectionName ?? '',
        })
      },
      [navigation, owner]
    )

    const onPressScan = (): void => {
      // in case we received a pending session from a previous scan after closing modal
      dispatch(removePendingSession())
      dispatch(
        openModal({ name: ModalName.WalletConnectScan, initialState: ScannerModalState.WalletQr })
      )
    }

    const renderItem = useCallback(
      ({ item }: ListRenderItemInfo<string | NFTItem>) => {
        return typeof item === 'string' ? (
          <Loader.NFT />
        ) : (
          <Box flex={1} justifyContent="flex-start" m="spacing4">
            <TouchableArea
              hapticFeedback
              activeOpacity={1}
              hapticStyle={ImpactFeedbackStyle.Light}
              onPress={(): void => onPressItem(item)}>
              <Box
                alignItems="center"
                aspectRatio={1}
                backgroundColor="backgroundOutline"
                borderRadius="rounded12"
                // eslint-disable-next-line react-native/no-inline-styles
                style={{ overflow: 'hidden' }}
                width="100%">
                <NFTViewer
                  imageDimensions={item.imageDimensions}
                  limitGIFSize={ESTIMATED_ITEM_SIZE}
                  maxHeight={MAX_NFT_IMAGE_SIZE}
                  placeholderContent={item.name || item.collectionName}
                  squareGridView={true}
                  uri={item.imageUrl ?? ''}
                />
              </Box>
            </TouchableArea>
          </Box>
        )
      },
      [onPressItem]
    )

    /**
     * If tab container is smaller than the approximate screen height, we need to manually add
     * padding so scroll works as intended since minHeight is not supported by FlashList in
     * `contentContainerStyle`. Padding is proportional to the number of rows the data items take up.
     */
    const footerPadding =
      nftDataItems.length < 6
        ? (ESTIMATED_ITEM_SIZE * (6 - nftDataItems.length)) / 2
        : FOOTER_HEIGHT

    const onRetry = useCallback(() => refetch(), [refetch])

    // Initial loading state or refetch
    if (isNonPollingRequestInFlight(networkStatus)) {
      return (
        <View
          style={[
            containerProps?.loadingContainerStyle,
            { paddingHorizontal: theme.spacing.spacing12 },
          ]}>
          <Loader.NFT repeat={6} />
        </View>
      )
    }

    if (isError(networkStatus, !!data)) {
      return (
        <Flex grow style={containerProps?.emptyContainerStyle}>
          <BaseCard.ErrorState
            description={t('Something went wrong.')}
            retryButtonLabel={t('Retry')}
            title={t('Couldnâ€™t load NFTs')}
            onRetry={onRetry}
          />
        </Flex>
      )
    }

    return nftDataItems.length === 0 ? (
      <Flex centered grow flex={1} style={containerProps?.emptyContainerStyle}>
        <BaseCard.EmptyState
          buttonLabel={t('Receive NFTs')}
          description={t('Transfer NFTs from another wallet to get started.')}
          icon={<NoNFTsIcon color={theme.colors.textSecondary} />}
          title={t('No NFTs yet')}
          onPress={onPressScan}
        />
      </Flex>
    ) : (
      <Flex grow px="spacing12">
        <AnimatedFlashList
          ref={ref}
          ListFooterComponent={
            // If not loading, we add a footer  to cover any possible space that is covered up by bottom tab bar
            networkStatus === NetworkStatus.fetchMore ? (
              <Loader.NFT repeat={4} />
            ) : (
              <Box height={footerPadding} />
            )
          }
          data={shouldAddInLoadingItem ? [...nftDataItems, LOADING_ITEM] : nftDataItems}
          estimatedItemSize={ESTIMATED_ITEM_SIZE}
          keyExtractor={keyExtractor}
          numColumns={2}
          renderItem={renderItem}
          showsVerticalScrollIndicator={false}
          onEndReached={onListEndReached}
          onEndReachedThreshold={PREFETCH_ITEMS_THRESHOLD}
          onScroll={scrollHandler}
          {...containerProps}
        />
      </Flex>
    )
  }
)
