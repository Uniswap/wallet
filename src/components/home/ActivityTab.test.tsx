import { MockedResponse } from '@apollo/client/testing'
import React from 'react'
import { ActivityTab } from 'src/components/home/ActivityTab'
import {
  Chain,
  TokenDocument,
  TokenQuery,
  TransactionListDocument,
  TransactionListQuery,
} from 'src/data/__generated__/types-and-hooks'
import { ACCOUNT_ADDRESS_ONE, mockWalletPreloadedState } from 'src/test/fixtures'
import { DaiAsset, Portfolios } from 'src/test/gqlFixtures'
import { act, render } from 'src/test/test-utils'
import { sleep } from 'src/utils/timing'

const TransactionListMock: MockedResponse<TransactionListQuery> = {
  request: {
    query: TransactionListDocument,
    variables: {
      address: ACCOUNT_ADDRESS_ONE,
    },
  },
  result: {
    data: {
      portfolios: Portfolios,
    },
  },
}

/**
 * Need to mock responses for the BalanceUpdate query that queries for this data
 * if needed.
 */

const TokenMock: MockedResponse<TokenQuery> = {
  request: {
    query: TokenDocument,
    variables: {
      chain: Chain.Ethereum,
      address: DaiAsset.address?.toLowerCase(),
    },
  },
  result: {
    data: {
      token: DaiAsset,
    },
  },
}

const TokenNullMock: MockedResponse<TokenQuery> = {
  request: {
    query: TokenDocument,
    variables: {
      chain: Chain.Ethereum,
      address: null,
    },
  },
  result: {
    data: {
      token: undefined,
    },
  },
}

describe('ActivityTab', () => {
  it('renders without error', async () => {
    const tokensTabScrollHandler = (): undefined => undefined

    const tree = render(
      <ActivityTab owner={ACCOUNT_ADDRESS_ONE} scrollHandler={tokensTabScrollHandler} />,
      {
        mocks: [TransactionListMock, TokenMock, TokenNullMock],
        preloadedState: mockWalletPreloadedState,
      }
    )

    // Loading
    expect(tree.toJSON()).toMatchSnapshot()

    // Render items
    await act(async () => {
      await sleep(1000) // Wait for the loading to complete
    })

    expect(tree.toJSON()).toMatchSnapshot()
  })
})
