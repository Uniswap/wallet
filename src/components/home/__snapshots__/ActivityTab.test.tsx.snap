// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityTab renders without error 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 1,
        "height": "100%",
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "center",
          "padding": 12,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "justifyContent": "center",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "marginBottom": 24,
            },
          ]
        }
      >
        <svg
          color="#505A78"
          data-file-name="SvgEmptyStateTransaction"
        />
        <View
          style={
            Array [
              Object {
                "bottom": -23,
                "display": "flex",
                "left": 5,
                "position": "absolute",
              },
            ]
          }
        >
          <svg
            color="#E8ECFB"
            data-file-name="SvgEmptyStateCoin"
            fill="#FC72FF"
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "flexGrow": 0,
              "flexShrink": 0,
              "height": 16,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "column",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          allowFontScaling={true}
          maxFontSizeMultiplier={1.2}
          style={
            Array [
              Object {
                "color": "#0D111C",
                "display": "flex",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 20,
                "textAlign": "center",
              },
            ]
          }
        >
          No activity yet
        </Text>
        <View
          style={
            Array [
              Object {
                "display": "flex",
                "flexGrow": 0,
                "flexShrink": 0,
                "height": 8,
              },
            ]
          }
        />
        <Text
          allowFontScaling={true}
          maxFontSizeMultiplier={1.4}
          style={
            Array [
              Object {
                "color": "#505A78",
                "display": "flex",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": 20,
                "textAlign": "center",
              },
            ]
          }
        >
          When this wallet makes transactions, theyâ€™ll appear here.
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flexGrow": 0,
            "flexShrink": 0,
            "height": 24,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "stretch",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "flex-start",
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`ActivityTab renders without error 2`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "stretch",
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 1,
        "justifyContent": "flex-start",
      },
      Object {
        "paddingHorizontal": 24,
      },
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      ListFooterComponent={
        <ForwardRef
          height={184}
        />
      }
      applyWindowCorrection={[Function]}
      canChangeSize={true}
      contentContainerStyle={
        Object {
          "backgroundColor": undefined,
          "minHeight": 1,
          "minWidth": 1,
        }
      }
      contentHeight={0}
      contentWidth={0}
      data={
        Array [
          Object {
            "itemType": "HEADER",
            "title": "February 1990",
          },
          Object {
            "addedTime": 634722508000,
            "chainId": 1,
            "from": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
            "hash": "0xaef17b4b697be91ab3d7eca4b92fa53efeb2da6b",
            "id": "0xaef17b4b697be91ab3d7eca4b92fa53efeb2da6b",
            "options": Object {
              "request": Object {},
            },
            "status": "confirmed",
            "typeInfo": Object {
              "approvalAmount": "1",
              "spender": "0x1234567890123456789012345678901234567890",
              "tokenAddress": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
              "type": "approve",
            },
          },
          Object {
            "addedTime": 634722508000,
            "chainId": 1,
            "from": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
            "hash": "0x8123abfa76b2bd459df69d4cffefde0e51b81cf2",
            "id": "0x8123abfa76b2bd459df69d4cffefde0e51b81cf2",
            "options": Object {
              "request": Object {},
            },
            "status": "confirmed",
            "typeInfo": Object {
              "expectedOutputCurrencyAmountRaw": "1000000000000000000",
              "inputCurrencyAmountRaw": "1000000000000000000",
              "inputCurrencyId": "1-0x6B175474E89094C44Da98b954EedeAC495271d0F",
              "minimumOutputCurrencyAmountRaw": "1000000000000000000",
              "outputCurrencyId": "1-0x6B175474E89094C44Da98b954EedeAC495271d0F",
              "tradeType": 0,
              "transactedUSDValue": 1,
              "type": "swap",
            },
          },
        ]
      }
      dataProvider={
        DataProvider {
          "_data": Array [
            Object {
              "itemType": "HEADER",
              "title": "February 1990",
            },
            Object {
              "addedTime": 634722508000,
              "chainId": 1,
              "from": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
              "hash": "0xaef17b4b697be91ab3d7eca4b92fa53efeb2da6b",
              "id": "0xaef17b4b697be91ab3d7eca4b92fa53efeb2da6b",
              "options": Object {
                "request": Object {},
              },
              "status": "confirmed",
              "typeInfo": Object {
                "approvalAmount": "1",
                "spender": "0x1234567890123456789012345678901234567890",
                "tokenAddress": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
                "type": "approve",
              },
            },
            Object {
              "addedTime": 634722508000,
              "chainId": 1,
              "from": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
              "hash": "0x8123abfa76b2bd459df69d4cffefde0e51b81cf2",
              "id": "0x8123abfa76b2bd459df69d4cffefde0e51b81cf2",
              "options": Object {
                "request": Object {},
              },
              "status": "confirmed",
              "typeInfo": Object {
                "expectedOutputCurrencyAmountRaw": "1000000000000000000",
                "inputCurrencyAmountRaw": "1000000000000000000",
                "inputCurrencyId": "1-0x6B175474E89094C44Da98b954EedeAC495271d0F",
                "minimumOutputCurrencyAmountRaw": "1000000000000000000",
                "outputCurrencyId": "1-0x6B175474E89094C44Da98b954EedeAC495271d0F",
                "tradeType": 0,
                "transactedUSDValue": 1,
                "type": "swap",
              },
            },
          ],
          "_firstIndexToProcess": 0,
          "_hasStableIds": true,
          "_requiresDataChangeHandling": false,
          "_size": 3,
          "getStableId": [Function],
          "rowHasChanged": [Function],
        }
      }
      disableRecycling={false}
      estimatedItemSize={92}
      extendedState={Object {}}
      externalScrollView={[Function]}
      finalRenderAheadOffset={250}
      forceNonDeterministicRendering={true}
      horizontal={false}
      initialOffset={0}
      initialRenderIndex={0}
      isHorizontal={false}
      keyExtractor={[Function]}
      layoutProvider={
        GridLayoutProviderWithProps {
          "_acceptableRelayoutDelta": 1,
          "_getHeightOrWidth": [Function],
          "_getLayoutTypeForIndex": [Function],
          "_getSpan": [Function],
          "_maxSpan": 1,
          "_setLayoutForType": [Function],
          "_tempDim": Object {
            "height": 0,
            "width": 0,
          },
          "averageWindow": AverageWindow {
            "currentAverage": 92,
            "currentCount": 1,
            "inputValues": Array [
              92,
            ],
            "nextIndex": 1,
          },
          "defaultEstimatedItemSize": 100,
          "layoutObject": Object {
            "size": undefined,
            "span": undefined,
          },
          "props": Object {
            "ListFooterComponent": <ForwardRef
              height={184}
            />,
            "data": Array [
              Object {
                "itemType": "HEADER",
                "title": "February 1990",
              },
              Object {
                "addedTime": 634722508000,
                "chainId": 1,
                "from": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
                "hash": "0xaef17b4b697be91ab3d7eca4b92fa53efeb2da6b",
                "id": "0xaef17b4b697be91ab3d7eca4b92fa53efeb2da6b",
                "options": Object {
                  "request": Object {},
                },
                "status": "confirmed",
                "typeInfo": Object {
                  "approvalAmount": "1",
                  "spender": "0x1234567890123456789012345678901234567890",
                  "tokenAddress": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
                  "type": "approve",
                },
              },
              Object {
                "addedTime": 634722508000,
                "chainId": 1,
                "from": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
                "hash": "0x8123abfa76b2bd459df69d4cffefde0e51b81cf2",
                "id": "0x8123abfa76b2bd459df69d4cffefde0e51b81cf2",
                "options": Object {
                  "request": Object {},
                },
                "status": "confirmed",
                "typeInfo": Object {
                  "expectedOutputCurrencyAmountRaw": "1000000000000000000",
                  "inputCurrencyAmountRaw": "1000000000000000000",
                  "inputCurrencyId": "1-0x6B175474E89094C44Da98b954EedeAC495271d0F",
                  "minimumOutputCurrencyAmountRaw": "1000000000000000000",
                  "outputCurrencyId": "1-0x6B175474E89094C44Da98b954EedeAC495271d0F",
                  "tradeType": 0,
                  "transactedUSDValue": 1,
                  "type": "swap",
                },
              },
            ],
            "estimatedItemSize": 92,
            "keyExtractor": [Function],
            "numColumns": 1,
            "onScroll": [Function],
            "renderItem": [Function],
            "showsVerticalScrollIndicator": false,
          },
          "shouldRefreshWithAnchoring": true,
        }
      }
      maxRenderAhead={750}
      numColumns={1}
      onEndReached={[Function]}
      onEndReachedThreshold={0}
      onEndReachedThresholdRelative={0}
      onItemLayout={[Function]}
      onLayout={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onSizeChanged={[Function]}
      onVisibleIndicesChanged={[Function]}
      removeClippedSubviews={false}
      renderAheadOffset={0}
      renderAheadStep={250}
      renderContentContainer={[Function]}
      renderItem={[Function]}
      renderItemContainer={[Function]}
      rowRenderer={[Function]}
      scrollEventThrottle={16}
      scrollThrottle={16}
      scrollViewProps={
        Object {
          "contentContainerStyle": Object {
            "backgroundColor": undefined,
            "minHeight": 1,
            "minWidth": 1,
          },
          "onLayout": [Function],
          "onScrollBeginDrag": [Function],
          "refreshControl": undefined,
          "style": Object {
            "minHeight": 1,
            "minWidth": 1,
          },
        }
      }
      showsVerticalScrollIndicator={false}
      style={
        Object {
          "minHeight": 1,
          "minWidth": 1,
        }
      }
      suppressBoundedSizeException={true}
      windowCorrectionConfig={
        Object {
          "applyToInitialOffset": true,
          "applyToItemScroll": true,
          "value": Object {
            "endCorrection": 0,
            "startCorrection": 0,
            "windowShift": -0,
          },
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "flexDirection": "column",
            }
          }
        >
          <AutoLayoutView
            enableInstrumentation={false}
            horizontal={false}
            onBlankAreaEvent={[Function]}
            onLayout={[Function]}
            renderAheadOffset={0}
            scrollOffset={0}
            style={
              Object {
                "height": 0,
                "width": 0,
              }
            }
            windowSize={0}
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
